lib/net.c

```diff
@@ -82,6 +75,115 @@ static struct rte_eth_conf gatekeeper_port_conf = {
 	},
 };
 
+/*
+ * @dst_ip, @src_port and @dst_port must be in big endian.
+ * By specifying the tuple (proto, src_port, dst_port),
+ * it can filter both IPv4 and IPv6 addresses.
+ */
+int
+ntuple_filter_add(uint8_t portid, uint32_t dst_ip,
+	uint16_t src_port, uint16_t dst_port, uint16_t queue_id)
+{
+	int ret = 0;
+	struct rte_eth_ntuple_filter filter_v4 = {
+		.flags = RTE_5TUPLE_FLAGS,
+		.dst_ip = dst_ip,
+		.dst_ip_mask = UINT32_MAX,
+		.src_ip = 0,
+		.src_ip_mask = 0,
+		.dst_port = dst_port,
+		.dst_port_mask = UINT16_MAX,
+		.src_port = src_port,
+		.src_port_mask = UINT16_MAX,
+		.proto = IPPROTO_UDP,
+		.proto_mask = UINT8_MAX,
+		.tcp_flags = 0,
+		.priority = 1,
+		.queue = queue_id,
+	};
+
+	struct rte_eth_ntuple_filter filter_v6 = {
+		.flags = RTE_5TUPLE_FLAGS,
+		.dst_ip = 0,
+		.dst_ip_mask = 0,
+		.src_ip = 0,
+		.src_ip_mask = 0,
+		.dst_port = dst_port,
+		.dst_port_mask = UINT16_MAX,
+		.src_port = src_port,
+		.src_port_mask = UINT16_MAX,
+		.proto = IPPROTO_UDP,
+		.proto_mask = UINT8_MAX,
+		.tcp_flags = 0,
+		.priority = 1,
+		.queue = queue_id,
+	};
+
+	ret = rte_eth_dev_filter_supported(portid, RTE_ETH_FILTER_NTUPLE);
+	if (ret < 0) {
+		RTE_LOG(ERR, PORT,
+			"Ntuple filter is not supported on port %hhu.\n",
+			portid);
+		ret = -1;
+		goto out;
+	}
+
+	if (dst_ip != 0) {
+		ret = rte_eth_dev_filter_ctrl(portid,
+			RTE_ETH_FILTER_NTUPLE,
+			RTE_ETH_FILTER_ADD,
+			&filter_v4);
+		if (ret == -ENOTSUP) {
+			RTE_LOG(ERR, PORT,
+				"Hardware doesn't support adding a ntuple filter on port %hhu!\n",
+				portid);
+			ret = -1;
+			goto out;
+		} else if (ret == -ENODEV) {
+			RTE_LOG(ERR, PORT,
+				"Port %hhu is invalid!\n",
+				portid);
+			ret = -1;
+			goto out;
+		} else if (ret != 0) {
+			RTE_LOG(ERR, PORT,
+				"Other errors that depend on the specific operations implementation on port %hhu!\n",
+				portid);
+			ret = -1;
+			goto out;
+		}
+	}
+
+	ret = rte_eth_dev_filter_ctrl(portid,
+		RTE_ETH_FILTER_NTUPLE,
+		RTE_ETH_FILTER_ADD,
+		&filter_v6);
+	if (ret == -ENOTSUP) {
+		RTE_LOG(ERR, PORT,
+			"Hardware doesn't support adding a ntuple filter on port %hhu!\n",
```

Comment:
In this and in the following two log entries, please mention "IPv6 ntuple filter" to be clear what failed.